File: dsa/challenges/multi_bracket_validation/multi_bracket_validation.py
from dsa.data_structures.stack_and_queues.stack_and_queues import Node, Stack

def multi_bracket_validation(input_string):
    br_dict = {'}':'{', ')':"(", ']':"["}
    if input_string == "":
        return False
    _stack = Stack()
    _stack_length = 0
    for chr in input_string:
        if chr in br_dict:
            if br_dict[chr]==_stack.peek():
                _stack.pop()
            else:
                return False
        elif chr in br_dict.values():
            _stack.push(chr)
            _stack_length += 1
    if _stack_length == 0:
        return "no brackets in string"
    elif _stack.isEmpty():
        return True
    return False


def matching_brackets(string):
    bracket = Stack()
    bracket_dict = {"(":")","{":"}", "[":"]" }
    for char in string:
        if char in bracket_dict:
            bracket.push(char)
        elif char in [")", "}", "]"]:
            if bracket.isEmpty():
                return False
            if bracket_dict[bracket.peek()]==char:
                bracket.pop()
            else:
                return False
    if not bracket.isEmpty():
        return False
    return True


if __name__ == "__main__":
    # input_string = "jjkjklj"
    input_string = "abcds{([])}"
    print(multi_bracket_validation(input_string))

File: dsa/java_challenges/src/main/java/java_challenges/utilities/MultiBracketValidation.java
package java_challenges.utilities;

import data_structures.linkedList.LinkedList;

import java.util.EmptyStackException;
import java.util.HashMap;

public class MultiBracketValidation {
    //create Node class with char
    protected static class Node {
        char value;
        Node next;
        public Node (char val){
            this.value = val;
            this.next = null;
        }
        public Node (char val, Node next){
            this.value = val;
            this.next = next;
        }
    }
    //create a stack class
    protected static class Stack {
        Node top = null;
        public boolean isEmpty(){
            return top==null;
        }
        public char peek () throws EmptyStackException {
            if (isEmpty()) {throw new EmptyStackException();}
            return top.value;
        }
        public void push(char val){
            Node newNode = new Node(val);
            if (isEmpty()){
                top = newNode;
                return;
            }
            top = new Node(val, top);
        }
        public char pop() throws EmptyStackException {
            if (top == null){
                throw new EmptyStackException();
            }
            Node temp = top;
            top = top.next;
            return temp.value;
        }
    }

    //make a stack and then check each bracket and then close it and pop it off
    // and if it is empty then we will return true
    public static boolean multiBracketValidationWithString (String inputString){
        System.out.println("this is the inputString "+inputString);
        //create a hashmap of the brackets with O1 look up time
        // instantiate a stack
        // iterate over the string and then add to the stack as we see
        HashMap<Character, Character> brackets = new HashMap<>();
        brackets.put('{','}');
        brackets.put('(',')');
        brackets.put('[',']');
        System.out.println("this is the hashmap  "+ brackets);
        Stack bracketStack= new Stack();
        int counter = 0;
        //iterate over the string and get the contents of the string
        // check if it matches the keys in the brackets and then ad it to the stack
        for (int i = 0; i< inputString.length(); i++){
            if (brackets.containsKey(inputString.charAt(i))){
                bracketStack.push(brackets.get(inputString.charAt(i)));
                counter ++;
                //check if the character is a closing bracket
            } else if (brackets.containsValue(inputString.charAt(i))){
                if (bracketStack.isEmpty()){return false;}
                //check if that character is in the Stack
                if (bracketStack.peek() == inputString.charAt(i)) {
                    bracketStack.pop();
                }
            }
        }
        if (bracketStack.isEmpty() && counter!=0){
            return true;
        }
        return false;
    }
}
